esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py

```diff
@@ -0,0 +1,208 @@
+        use_tgt_mask: bool | np.ndarray = True,
+        src_resolution: int | None = None,
+        tgt_resolution: int | None = None,
+        tgt_location: Literal['face', 'node'] | None = None,
+    ) -> None:
+        if method not in METHODS:
+            raise ValueError(
+                "`method` should be one of 'bilinear', 'conservative', or "
+                "'nearest'")
+
+        self.kwargs: dict[str, Any] = {
+            'method': method,
+            'use_src_mask': use_src_mask,
+            'use_tgt_mask': use_tgt_mask,
+            'tgt_location': tgt_location,
+        }
+        if method == 'nearest':
+            if mdtol is not None:
+                raise TypeError(
+                    "`mdol` can only be specified when `method='bilinear'` "
+                    "or `method='conservative'`")
+        else:
+            self.kwargs['mdtol'] = mdtol
+        if method == 'conservative':
+            self.kwargs['src_resolution'] = src_resolution
+            self.kwargs['tgt_resolution'] = tgt_resolution
+        elif src_resolution is not None:
+            raise TypeError("`src_resolution` can only be specified when "
+                            "`method='conservative'`")
+        elif tgt_resolution is not None:
+            raise TypeError("`tgt_resolution` can only be specified when "
+                            "`method='conservative'`")
+
+    def __repr__(self) -> str:
+        """Return string representation of class."""
+        kwargs_str = ", ".join(f"{k}={repr(v)}"
+                               for k, v in self.kwargs.items())
+        return f'{self.__class__.__name__}({kwargs_str})'
+
+    @staticmethod
+    def _get_mask(cube: iris.cube.Cube) -> np.ndarray:
+        """Read the mask from the cube data.
+
+        If the cube has a vertical dimension, the mask will consist of
+        those points which are masked in all vertical levels.
+
+        This function assumes that the mask is constant in dimensions
+        that are not horizontal or vertical.
+        """
+        horizontal_dims = _get_horizontal_dims(cube)
+        try:
+            vertical_coord = cube.coord(axis="z", dim_coords=True)
+        except iris.exceptions.CoordinateNotFoundError:
+            vertical_coord = None
+
+        data = cube.core_data()
+        if vertical_coord is None:
+            slices = tuple(
+                slice(None) if i in horizontal_dims else 0
+                for i in range(cube.ndim))
+            mask = da.ma.getmaskarray(data[slices])
+        else:
+            vertical_dim = cube.coord_dims(vertical_coord)[0]
+            slices = tuple(
+                slice(None) if i in horizontal_dims + (vertical_dim, ) else 0
+                for i in range(cube.ndim))
+            mask = da.ma.getmaskarray(data[slices])
+            mask_vertical_dim = sum(i < vertical_dim for i in horizontal_dims)
+            mask = mask.all(axis=mask_vertical_dim)
+        return mask
```

Comment:
Why do we need to explicitly calculate the mask here? From the [docs](https://iris-esmf-regrid.readthedocs.io/en/stable/_api_generated/esmf_regrid.schemes.html#module-esmf_regrid.schemes), it looks like `use_src_mask` and `use_tgt_mask` can only be booleans. So why don't we just use `True/False` here?

It's also dangerous to derive this only from the first time step, as the mask might be time-dependent (e.g. for satellite data).

Reply:
It needs to be an array for data that contains a vertical dimension. This is allowed, as documented [here](https://iris-esmf-regrid.readthedocs.io/en/stable/_api_generated/esmf_regrid.schemes.html#esmf_regrid.schemes.ESMFAreaWeightedRegridder) and enforced by this code: https://github.com/SciTools-incubator/iris-esmf-regrid/blob/1772591d2ee1dbab06312e087ff694abe370cb72/esmf_regrid/schemes.py#L37-L76.

The mask seems mostly important to hide discontinuities in the grid. As long as the latitude and longitude coordinates are not dependent on time or vertical level, the mask computed here should be valid. It could mask out too much though, if fewer points are masked at a later timestep. However, computing the mask over all timesteps is computationally expensive as it requires loading the entire dataset from disk.

Reply:
Are we currently supporting satellite datasets with time-varying masks?

Reply:
> Are we currently supporting satellite datasets with time-varying masks?

Yes, we have some in our data pool. It's also quite common that a mask is height-dependent, i.e., for lower levels the regions with high altitude are masked.

I assume the mask you can pass to the functions can only be 2D? This is really not optimal...it would be much simpler if iris-esmf supported masks with the same shape as the input data.

Reply:
>  It's also quite common that a mask is height-dependent, i.e., for lower levels the regions with high altitude are masked.

The current code already takes care of that. I'll add something for the case where the mask is time varying.

> I assume the mask you can pass to the functions can only be 2D?

The mask is used to compute the regridding weights. Therefore the mask that is required describes the grid, so its dimensions need to match those of the grid.

Reply:
I added the option to specify along which axes the mask varies using the `collapse_src_mask_along_axes` and `collapse_tgt_mask_along_axis` arguments, see [here](https://esmvaltool--2457.org.readthedocs.build/projects/ESMValCore/en/2457/api/esmvalcore.regridding_schemes.html#esmvalcore.preprocessor.regrid_schemes.IrisESMFRegrid) for the documentation. Does it look better now?